File: plox/src/Lox/Parser.php
<?php

namespace Nkoll\Plox\Lox;

use Nkoll\Plox\Lox\Error\ParserError;
use Nkoll\Plox\Lox\Expr\AssignExpr;
use Nkoll\Plox\Lox\Expr\BinaryExpr;
use Nkoll\Plox\Lox\Expr\CallExpr;
use Nkoll\Plox\Lox\Expr\Expr;
use Nkoll\Plox\Lox\Expr\GroupingExpr;
use Nkoll\Plox\Lox\Expr\LiteralExpr;
use Nkoll\Plox\Lox\Expr\LogicalExpr;
use Nkoll\Plox\Lox\Expr\UnaryExpr;
use Nkoll\Plox\Lox\Expr\VariableExpr;
use Nkoll\Plox\Lox\Stmt\BlockStmt;
use Nkoll\Plox\Lox\Stmt\ExpressionStmt;
use Nkoll\Plox\Lox\Stmt\FunctionStmt;
use Nkoll\Plox\Lox\Stmt\IfStmt;
use Nkoll\Plox\Lox\Stmt\PrintStmt;
use Nkoll\Plox\Lox\Stmt\ReturnStmt;
use Nkoll\Plox\Lox\Stmt\Stmt;
use Nkoll\Plox\Lox\Stmt\VarStmt;
use Nkoll\Plox\Lox\Stmt\WhileStmt;
use Nkoll\Plox\PloxCommand;

class Parser
{
    private array $tokens;
    private int $current = 0;

    public function __construct(array $tokens)
    {
        $this->tokens = $tokens;
    }

    /** @return ?Stmt[]  */
    public function parse(): ?array
    {
        try {
            $stmts = [];
            while (!$this->isAtEnd()) {
                $stmts[] = $this->declaration();
            }

            return $stmts;
        } catch (ParserError) {
            return null;
        }
    }

    private function declaration(): ?Stmt
    {
        try {
            if ($this->match(TokenType::FUN)) {
                return $this->function('function');
            }
            if ($this->match(TokenType::VAR)) {
                return $this->varDeclaration();
            }

            return $this->statement();
        } catch (ParserError) {
            $this->synchronize();

            return null;
        }
    }

    private function varDeclaration(): Stmt
    {
        $name = $this->consume(TokenType::IDENTIFIER, 'Expect variable name.');

        $initializer = null;
        if ($this->match(TokenType::EQUAL)) {
            $initializer = $this->expression();
        }

        $this->consume(TokenType::SEMICOLON, "Expect ';' after variable declaration.");

        return new VarStmt($name, $initializer);
    }

    private function statement(): Stmt
    {
        if ($this->match(TokenType::IF)) {
            return $this->ifStatement();
        }

        if ($this->match(TokenType::WHILE)) {
            return $this->whileStatement();
        }

        if ($this->match(TokenType::FOR)) {
            return $this->forStatement();
        }

        if ($this->match(TokenType::PRINT)) {
            return $this->printStatement();
        }

        if ($this->match(TokenType::RETURN)) {
            return $this->returnStatement();
        }

        if ($this->match(TokenType::LEFT_BRACE)) {
            return new BlockStmt($this->block());
        }

        return $this->expressionStatement();
    }

    private function ifStatement(): Stmt
    {
        $this->consume(TokenType::LEFT_PAREN, "Expect '(' after 'if'.");
        $condition = $this->expression();
        $this->consume(TokenType::RIGHT_PAREN, "Expect ')' after condition.");

        $thenBranch = $this->statement();
        $elseBranch = null;
        if ($this->match(TokenType::ELSE)) {
            $elseBranch = $this->statement();
        }

        return new IfStmt($condition, $thenBranch, $elseBranch);
    }

    private function whileStatement(): Stmt
    {
        $this->consume(TokenType::LEFT_PAREN, "Expect '(' after 'while'.");
        $condition = $this->expression();
        $this->consume(TokenType::RIGHT_PAREN, "Expect ')' after condition.");
        $body = $this->statement();

        return new WhileStmt($condition, $body);
    }

    private function forStatement(): Stmt
    {
        $this->consume(TokenType::LEFT_PAREN, "Expect '(' after 'for'.");

        $initializer = null;
        if ($this->match(TokenType::SEMICOLON)) {
            $initializer = null; // skipped
        } elseif ($this->match(TokenType::VAR)) {
            $initializer = $this->varDeclaration();
        } else {
            $initializer = $this->expressionStatement();
        }

        $condition = null;
        if (!$this->check(TokenType::SEMICOLON)) {
            $condition = $this->expression();
        }
        $this->consume(TokenType::SEMICOLON, "Expect ';' after loop condition.");

        $increment = null;
        if (!$this->check(TokenType::RIGHT_PAREN)) {
            $increment = $this->expression();
        }
        $this->consume(TokenType::RIGHT_PAREN, "Expect ')' after for clauses.");
        $body = $this->statement();

        if ($increment) {
            $body = new BlockStmt([$body, new ExpressionStmt($increment)]);
        }

        if (!$condition) {
            $condition = new LiteralExpr(true);
        }
        $body = new WhileStmt($condition, $body);

        if ($initializer) {
            $body = new BlockStmt([$initializer, $body]);
        }

        return $body;
    }

    private function printStatement(): Stmt
    {
        $expr = $this->expression();
        $this->consume(TokenType::SEMICOLON, "Expect ';' after value.");

        return new PrintStmt($expr);
    }

    private function returnStatement(): Stmt {
        $keyword = $this->previous();

        $value = null;
        if (!$this->check(TokenType::SEMICOLON)) {
            $value = $this->expression();
        }

        $this->consume(TokenType::SEMICOLON, "Expected ';' after return value.");
        return new ReturnStmt($keyword, $value);
    }

    private function expressionStatement(): Stmt
    {
        $expr = $this->expression();
        $this->consume(TokenType::SEMICOLON, "Expect ';' after expression.");

        return new ExpressionStmt($expr);
    }

    private function function(string $kind): Stmt
    {
        $name = $this->consume(TokenType::IDENTIFIER, "Expect $kind name.");
        $this->consume(TokenType::LEFT_PAREN, "Expect '(' after $kind name.");
        $params = [];
        if (!$this->check(TokenType::RIGHT_PAREN)) {
            do {
                if (count($params) >= 255) {
                    $this->error($this->peek(), "Can't have more than 255 parameters.");
                }

                $params[] = $this->consume(TokenType::IDENTIFIER, 'Expect parameter name.');
            } while ($this->match(TokenType::COMMA));
        }

        $this->consume(TokenType::RIGHT_PAREN, "Expect ')' after parameters.");

        $this->consume(TokenType::LEFT_BRACE, "Expect '{' before $kind body.");
        $body = $this->block();

        return new FunctionStmt($name, $params, $body);
    }

    /** @return Stmt[]  */
    private function block(): array
    {
        $stmts = [];

        while (!$this->check(TokenType::RIGHT_BRACE) && !$this->isAtEnd()) {
            $stmts[] = $this->declaration();
        }

        $this->consume(TokenType::RIGHT_BRACE, "Expect '}' after block.");

        return $stmts;
    }

    private function expression(): Expr
    {
        return $this->assignment();
    }

    private function assignment(): Expr
    {
        $expr = $this->or();

        if ($this->match(TokenType::EQUAL)) {
            $equals = $this->previous();
            $value = $this->assignment();

            if ($expr instanceof VariableExpr) {
                $name = $expr->name;

                return new AssignExpr($name, $value);
            }

            $this->error($equals, 'Invalid assignment target.');
        }

        return $expr;
    }

    private function or(): Expr
    {
        $expr = $this->and();

        while ($this->match(TokenType::OR)) {
            $op = $this->previous();
            $right = $this->and();
            $expr = new LogicalExpr($expr, $op, $right);
        }

        return $expr;
    }

    private function and(): Expr
    {
        $expr = $this->equality();

        while ($this->match(TokenType::AND)) {
            $op = $this->previous();
            $right = $this->equality();
            $expr = new LogicalExpr($expr, $op, $right);
        }

        return $expr;
    }

    private function equality(): Expr
    {
        $expr = $this->comparison();

        while ($this->match(TokenType::BANG_EQUAL, TokenType::EQUAL_EQUAL)) {
            $op = $this->previous();
            $right = $this->comparison();
            $expr = new BinaryExpr($expr, $op, $right);
        }

        return $expr;
    }

    private function comparison(): Expr
    {
        $expr = $this->term();

        while ($this->match(TokenType::GREATER, TokenType::GREATER_EQUAL, TokenType::LESS, TokenType::LESS_EQUAL)) {
            $op = $this->previous();
            $right = $this->term();
            $expr = new BinaryExpr($expr, $op, $right);
        }

        return $expr;
    }

    private function term(): Expr
    {
        $expr = $this->factor();

        while ($this->match(TokenType::PLUS, TokenType::MINUS)) {
            $op = $this->previous();
            $right = $this->factor();
            $expr = new BinaryExpr($expr, $op, $right);
        }

        return $expr;
    }

    private function factor(): Expr
    {
        $expr = $this->unary();

        while ($this->match(TokenType::STAR, TokenType::SLASH)) {
            $op = $this->previous();
            $right = $this->unary();
            $expr = new BinaryExpr($expr, $op, $right);
        }

        return $expr;
    }

    private function unary(): Expr
    {
        if ($this->match(TokenType::BANG, TokenType::MINUS)) {
            return new UnaryExpr($this->previous(), $this->unary());
        }

        return $this->call();
    }

    private function finishCall(Expr $callee): Expr
    {
        $args = [];

        if (!$this->check(TokenType::RIGHT_PAREN)) {
            do {
                if (count($args) >= 255) {
                    $this->error($this->peek(), "Can't have more than 255 arguments.");
                }
                $args[] = $this->expression();
            } while ($this->match(TokenType::COMMA));
        }

        $paren = $this->consume(TokenType::RIGHT_PAREN, "Expect ')' after arguments.");

        return new CallExpr($callee, $paren, $args);
    }

    private function call(): Expr
    {
        $expr = $this->primary();

        while (true) {
            if ($this->match(TokenType::LEFT_PAREN)) {
                $expr = $this->finishCall($expr);
            } else {
                break;
            }
        }

        return $expr;
    }

    private function primary(): Expr
    {
        if ($this->match(TokenType::FALSE)) {
            return new LiteralExpr(false);
        }
        if ($this->match(TokenType::TRUE)) {
            return new LiteralExpr(true);
        }
        if ($this->match(TokenType::NIL)) {
            return new LiteralExpr(null);
        }

        if ($this->match(TokenType::NUMBER, TokenType::STRING)) {
            return new LiteralExpr($this->previous()->literal);
        }

        if ($this->match(TokenType::IDENTIFIER)) {
            return new VariableExpr($this->previous());
        }

        if ($this->match(TokenType::LEFT_PAREN)) {
            $expr = $this->expression();
            $this->consume(TokenType::RIGHT_PAREN, "Expect ')' after expression.");

            return new GroupingExpr($expr);
        }

        throw $this->error($this->peek(), 'Expect expression.');
    }

    private function match(TokenType ...$types): bool
    {
        foreach ($types as $type) {
            if ($this->check($type)) {
                $this->advance();

                return true;
            }
        }

        return false;
    }

    private function consume(TokenType $type, string $msg): Token
    {
        if ($this->check($type)) {
            return $this->advance();
        }

        throw $this->error($this->peek(), $msg);
    }

    private function check(TokenType $type): bool
    {
        if ($this->isAtEnd()) {
            return false;
        }

        return $this->peek()->type === $type;
    }

    private function advance(): Token
    {
        if (!$this->isAtEnd()) {
            ++$this->current;
        }

        return $this->previous();
    }

    private function isAtEnd(): bool
    {
        return TokenType::EOF === $this->peek()->type;
    }

    private function peek(): Token
    {
        return $this->tokens[$this->current];
    }

    private function previous(): Token
    {
        return $this->tokens[$this->current - 1];
    }

    private function error(Token $token, string $msg): ParserError
    {
        PloxCommand::errorToken($token, $msg);

        return new ParserError();
    }

    private function synchronize()
    {
        $this->advance();

        while (!$this->isAtEnd()) {
            if (TokenType::SEMICOLON === $this->previous()->type) {
                return;
            }

            switch ($this->peek()->type) {
                case TokenType::CLASS_KEYWORD:
                case TokenType::FUN:
                case TokenType::VAR:
                case TokenType::FOR:
                case TokenType::IF:
                case TokenType::WHILE:
                case TokenType::PRINT:
                case TokenType::RETURN:
                    return;
            }

            $this->advance();
        }
    }
}

File: zox/src/compiler/Parser.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const StdErr = std.io.getStdErr();

const Scanner = @import("Scanner.zig");
const Token = @import("Token.zig");
const ast = @import("ast.zig");

const Parser = @This();

const Error = error{
    UnexpectedToken,
    MissingExpression,
    WrongNumberFormat,
    CouldNotGenerateNode,
    InvalidAssignment,
    TooManyArguments,
    TooManyParameters,
};

scanner: *Scanner,
current: Token,
previous: ?Token,
alloc: std.heap.ArenaAllocator,
has_error: bool,

pub fn init(alloc: Allocator, scanner: *Scanner) Parser {
    return .{
        .scanner = scanner,
        .current = undefined,
        .previous = null,
        .alloc = std.heap.ArenaAllocator.init(alloc),
        .has_error = false,
    };
}

/// All created nodes will be deinited, make sure to copy them if longer lifetime is needed
pub fn deinit(self: *Parser) void {
    self.alloc.deinit();
}

pub fn parse(self: *Parser) ?[]*ast.Stmt {
    self.current = self.scanner.nextToken(); // Priming the Parser

    var statements = std.ArrayList(*ast.Stmt).init(self.alloc.allocator());
    defer statements.deinit();

    while (!self.atEnd()) {
        const maybe_stmt = self.decl();
        if (maybe_stmt) |stmt| {
            statements.append(stmt) catch return null;
        }
    }

    if (self.has_error) {
        return null;
    }
    return statements.toOwnedSlice() catch null;
}

fn decl(self: *Parser) ?*ast.Stmt {
    return self.declaration() catch {
        self.sync();
        self.has_error = true;
        return null;
    };
}

fn declaration(self: *Parser) !*ast.Stmt {
    if (self.match(.VAR)) {
        return self.varDeclaration();
    }

    if (self.match(.FUN)) {
        return self.function();
    }

    return self.statement();
}

fn function(self: *Parser) !*ast.Stmt {
    try self.consume(.IDENTIFIER, "Expect function name");
    const name = self.previous.?;
    try self.consume(.@"(", "Expect '(' after function name");
    var parameters: std.ArrayList(Token) = .init(self.alloc.allocator());
    defer parameters.deinit();
    if (!self.check(.@")")) {
        while (true) {
            if (parameters.items.len > 255) {
                printError("A function can only expect a maximum of 255 parameters.", .{});
                return Error.TooManyParameters;
            }

            try self.consume(.IDENTIFIER, "Expect parameter name");
            try parameters.append(self.previous.?);

            if (!self.match(.@",")) break;
        }
    }

    try self.consume(.@")", "Expect ')' after parameters");
    try self.consume(.@"{", "Expect '{{' before function body");
    var body = try self.block();

    if (body.len == 0 or body[body.len - 1].* != .return_stmt) {
        const new_body = self.alloc.allocator().alloc(*ast.Stmt, body.len + 1) catch return Error.CouldNotGenerateNode;
        for (body, 0..) |stmt, i| {
            new_body[i] = stmt;
        }
        new_body[body.len] = ast.Stmt.newReturn(self.alloc.allocator(), null) catch return Error.CouldNotGenerateNode;
        body = new_body;
    }

    const params = try parameters.toOwnedSlice();
    return ast.Stmt.newFunction(self.alloc.allocator(), name, params, body);
}

fn varDeclaration(self: *Parser) !*ast.Stmt {
    try self.consume(.IDENTIFIER, "Expect variable name");
    const name = self.previous.?;

    const initializer: ?*ast.Expr = expr: {
        if (self.match(.@"=")) {
            break :expr try self.expression();
        }
        break :expr null;
    };

    try self.consume(.@";", "Expect ';' after variable declaration");

    return ast.Stmt.newVar(self.alloc.allocator(), name, initializer) catch Error.CouldNotGenerateNode;
}

fn statement(self: *Parser) Error!*ast.Stmt {
    if (self.match(.PRINT)) {
        return self.printStatement();
    }

    if (self.match(.RETURN)) {
        return self.returnStatement();
    }

    if (self.match(.WHILE)) {
        return self.whileStatement();
    }

    if (self.match(.IF)) {
        return self.ifStatement();
    }

    if (self.match(.FOR)) {
        return self.forStatement();
    }

    if (self.match(.@"{")) {
        const stmts = try self.block();
        return ast.Stmt.newBlock(self.alloc.allocator(), stmts) catch Error.CouldNotGenerateNode;
    }

    return self.expressionStatement();
}

fn block(self: *Parser) Error![]*ast.Stmt {
    var statements = std.ArrayList(*ast.Stmt).init(self.alloc.allocator());

    while (!self.atEnd() and !self.check(.@"}")) {
        const maybe_stmt = self.decl();
        if (maybe_stmt) |stmt| {
            statements.append(stmt) catch return Error.CouldNotGenerateNode;
        }
    }

    try self.consume(.@"}", "Expected '}}' after block");
    return statements.toOwnedSlice() catch return Error.CouldNotGenerateNode;
}

fn forStatement(self: *Parser) Error!*ast.Stmt {
    try self.consume(.@"(", "Expected '(' after for keyword");

    const initializer: ?*ast.Stmt = init: {
        if (self.match(.@";")) {
            break :init null;
        }

        if (self.match(.VAR)) {
            break :init try self.varDeclaration();
        }

        break :init try self.expressionStatement();
    };

    const condition: *ast.Expr = condition: {
        if (self.check(.@";")) {
            break :condition ast.Expr.newLiteral(self.alloc.allocator(), .{ .boolean = true }) catch return Error.CouldNotGenerateNode;
        }
        break :condition try self.expression();
    };

    try self.consume(.@";", "Expected ';' after loop condition");

    const increment: ?*ast.Stmt = inc: {
        if (self.check(.@")")) {
            break :inc null;
        }
        const expr = try self.expression();
        break :inc ast.Stmt.newExpression(self.alloc.allocator(), expr) catch return Error.CouldNotGenerateNode;
    };

    try self.consume(.@")", "Expected ')' after increment");
    const body = try self.statement();

    if (increment) |inc_expr| {
        body.block.stmts = self.alloc.allocator().realloc(body.block.stmts, body.block.stmts.len + 1) catch return Error.CouldNotGenerateNode;
        body.block.stmts[body.block.stmts.len - 1] = inc_expr;
    }

    const while_stmt = ast.Stmt.newWhile(self.alloc.allocator(), condition, body) catch return Error.CouldNotGenerateNode;

    if (initializer) |init_stmt| {
        const stmts = self.alloc.allocator().alloc(*ast.Stmt, 2) catch return Error.CouldNotGenerateNode;
        stmts[0] = init_stmt;
        stmts[1] = while_stmt;
        return ast.Stmt.newBlock(self.alloc.allocator(), stmts) catch Error.CouldNotGenerateNode;
    }

    return while_stmt;
}

fn printStatement(self: *Parser) Error!*ast.Stmt {
    const expr = try self.expression();
    try self.consume(.@";", "Expected ';' after expression");
    return ast.Stmt.newPrint(self.alloc.allocator(), expr) catch Error.CouldNotGenerateNode;
}

fn returnStatement(self: *Parser) Error!*ast.Stmt {
    if (self.match(.@";")) {
        return ast.Stmt.newReturn(self.alloc.allocator(), null) catch Error.CouldNotGenerateNode;
    }

    const expr = try self.expression();
    try self.consume(.@";", "Expected ';' after expression");
    return ast.Stmt.newReturn(self.alloc.allocator(), expr) catch Error.CouldNotGenerateNode;
}

fn whileStatement(self: *Parser) Error!*ast.Stmt {
    try self.consume(.@"(", "Expected '(' after while keyword");
    const condition = try self.expression();
    try self.consume(.@")", "Expected ')' after while condition");
    const stmt = try self.statement();
    return ast.Stmt.newWhile(self.alloc.allocator(), condition, stmt) catch Error.CouldNotGenerateNode;
}

fn ifStatement(self: *Parser) Error!*ast.Stmt {
    try self.consume(.@"(", "Expected '(' after if keyword");
    const condition = try self.expression();
    try self.consume(.@")", "Expected ')' after if condition");
    const stmt = try self.statement();
    if (self.match(.ELSE)) {
        const else_stmt = try self.statement();
        return ast.Stmt.newIf(self.alloc.allocator(), condition, stmt, else_stmt) catch Error.CouldNotGenerateNode;
    }
    return ast.Stmt.newIf(self.alloc.allocator(), condition, stmt, null) catch Error.CouldNotGenerateNode;
}

fn expressionStatement(self: *Parser) Error!*ast.Stmt {
    const expr = try self.expression();
    try self.consume(.@";", "Expected ';' after expression");
    return ast.Stmt.newExpression(self.alloc.allocator(), expr) catch Error.CouldNotGenerateNode;
}

fn expression(self: *Parser) Error!*ast.Expr {
    return self.assignment();
}

fn assignment(self: *Parser) Error!*ast.Expr {
    const expr = try self.logical_or();
    if (self.match(.@"=")) {
        if (expr.* == .variable) {
            const value = try self.expression();
            return ast.Expr.newAssignment(self.alloc.allocator(), expr.variable.name, value) catch Error.CouldNotGenerateNode;
        }

        printError("Invalid assignment target", .{});
        return Error.InvalidAssignment;
    }

    return expr;
}

fn logical_or(self: *Parser) Error!*ast.Expr {
    var left = try self.logical_and();
    while (self.match(.OR)) {
        const op = self.previous.?;
        const right = try self.logical_and();
        left = try self.logical(op, left, right);
    }

    return left;
}

fn logical_and(self: *Parser) Error!*ast.Expr {
    var left = try self.equality();
    while (self.match(.AND)) {
        const op = self.previous.?;
        const right = try self.equality();
        left = try self.logical(op, left, right);
    }

    return left;
}

fn equality(self: *Parser) Error!*ast.Expr {
    var left = try self.comparison();
    while (self.match(.@"==") or self.match(.@"!=")) {
        const op = self.previous.?;
        const right = try self.comparison();
        left = try self.binary(op, left, right);
    }

    return left;
}

fn comparison(self: *Parser) Error!*ast.Expr {
    var left = try self.term();
    while (self.match(.@"<") or self.match(.@"<=") or self.match(.@">") or self.match(.@">=")) {
        const op = self.previous.?;
        const right = try self.term();
        left = try self.binary(op, left, right);
    }
    return left;
}

fn term(self: *Parser) Error!*ast.Expr {
    var left = try self.factor();
    while (self.match(.@"+") or self.match(.@"-")) {
        const op = self.previous.?;
        const right = try self.factor();
        left = try self.binary(op, left, right);
    }

    return left;
}

fn factor(self: *Parser) Error!*ast.Expr {
    var left = try self.unary();
    while (self.match(.@"/") or self.match(.@"*")) {
        const op = self.previous.?;
        const right = try self.unary();
        left = try self.binary(op, left, right);
    }

    return left;
}

fn unary(self: *Parser) Error!*ast.Expr {
    if (self.match(.@"-") or self.match(.@"!")) {
        return ast.Expr.newUnary(self.alloc.allocator(), self.previous.?, try self.unary()) catch Error.CouldNotGenerateNode;
    }

    return self.call();
}

fn call(self: *Parser) Error!*ast.Expr {
    const value = try self.primary();

    if (!self.match(.@"(")) {
        return value;
    }

    if (self.match(.@")")) {
        return ast.Expr.newCall(self.alloc.allocator(), value, &.{}) catch Error.CouldNotGenerateNode;
    }

    var arguments: std.ArrayList(*ast.Expr) = .init(self.alloc.allocator());
    defer arguments.deinit();

    while (true) {
        if (arguments.items.len > 255) {
            printError("A function can accept a maximum of 255 arguments", .{});
            return Error.TooManyArguments;
        }
        const argument = try self.expression();
        arguments.append(argument) catch return Error.CouldNotGenerateNode;
        if (!self.match(.@",")) {
            break;
        }
    }

    try self.consume(.@")", "Expect ')' after arguments");
    const arguments_raw = arguments.toOwnedSlice() catch return Error.CouldNotGenerateNode;
    return ast.Expr.newCall(self.alloc.allocator(), value, arguments_raw) catch Error.CouldNotGenerateNode;
}

fn primary(self: *Parser) Error!*ast.Expr {
    if (self.match(.FALSE)) {
        return self.literal(.{ .boolean = false });
    }
    if (self.match(.TRUE)) {
        return self.literal(.{ .boolean = true });
    }
    if (self.match(.NIL)) {
        return self.literal(.nil);
    }
    if (self.match(.STRING)) {
        const prev_string = self.previousLexeme().?;
        const string = prev_string[1 .. prev_string.len - 1];
        return self.literal(.{ .string = string });
    }
    if (self.match(.NUMBER)) {
        const number = std.fmt.parseFloat(f64, self.previousLexeme().?) catch {
            printError("'{s}' is not a number", .{self.previousLexeme().?});
            return Error.WrongNumberFormat;
        };
        return self.literal(.{ .number = number });
    }
    if (self.match(.IDENTIFIER)) {
        return ast.Expr.newVariable(self.alloc.allocator(), self.previous.?) catch Error.CouldNotGenerateNode;
    }
    if (self.match(.@"(")) {
        const expr = try self.expression();
        try self.consume(.@")", "Expect ')' after expression");
        return expr;
    }

    printError("Expected Expression", .{});
    return Error.MissingExpression;
}

fn sync(self: *Parser) void {
    self.advance();

    while (!self.atEnd()) {
        if (self.previous != null and self.previous.?.type == .@";") {
            return;
        }

        switch (self.current.type) {
            .FUN,
            .VAR,
            .IF,
            .FOR,
            .WHILE,
            .RETURN,
            .PRINT,
            => return,
            else => self.advance(),
        }
    }
}

fn literal(self: *Parser, value: ast.Expr.Literal.Value) Error!*ast.Expr {
    return ast.Expr.newLiteral(self.alloc.allocator(), value) catch Error.CouldNotGenerateNode;
}

fn binary(self: *Parser, op: Token, left: *ast.Expr, right: *ast.Expr) Error!*ast.Expr {
    return ast.Expr.newBinary(self.alloc.allocator(), op, left, right) catch return Error.CouldNotGenerateNode;
}

fn logical(self: *Parser, op: Token, left: *ast.Expr, right: *ast.Expr) Error!*ast.Expr {
    return ast.Expr.newLogical(self.alloc.allocator(), op, left, right) catch return Error.CouldNotGenerateNode;
}

fn advance(self: *Parser) void {
    self.previous = self.current;
    self.current = self.scanner.nextToken();
}

fn check(self: *Parser, t_type: Token.Type) bool {
    return self.current.type == t_type;
}

fn match(self: *Parser, t_type: Token.Type) bool {
    if (self.check(t_type)) {
        self.advance();
        return true;
    }
    return false;
}

fn consume(self: *Parser, t_type: Token.Type, comptime msg: []const u8) Error!void {
    if (self.check(t_type)) {
        self.advance();
        return;
    }

    printError(msg, .{});
    return Error.UnexpectedToken;
}

fn atEnd(self: *Parser) bool {
    return self.check(.EOF);
}

fn currentLexeme(self: *Parser) []const u8 {
    return self.current.lexeme;
}

fn previousLexeme(self: *Parser) ?[]const u8 {
    if (self.previous) |prev| {
        return prev.lexeme;
    }
    return null;
}

fn printError(comptime format: []const u8, args: anytype) void {
    StdErr.writer().print(format, args) catch {};
    StdErr.writeAll("\n") catch {};
}
